src/action/timeline.rs

```diff
@@ -0,0 +1,567 @@
+use std::result::Result;
+use std::vec::Vec;
+
+use cfg_if::cfg_if;
+use sha2::{Digest, Sha256};
+use rrg_proto::{TimelineArgs, TimelineEntry, TimelineResult, DataBlob};
+
+use crate::gzchunked::{GzChunkedEncoder, GzChunkedCompression};
+use crate::session::{self, Session, Error, ParseError, MissingFieldError};
+
+/// A request type for the timeline action.
+pub struct Request {
+    root: PathBuf,
+}
+
+/// A newtype wrapper for SHA-256 chunk digest.
+#[derive(Debug, PartialEq, Clone)]
+struct ChunkDigest([u8; 32]);
+
+/// A response type for the timeline action (actual response).
+struct Response {
+    ids: Vec<ChunkDigest>,
+}
+
+/// A response type for the timeline action (transfer store chunks).
+struct ChunkResponse {
+    data: Vec<u8>,
+}
+
+/// An object for recursively traversing filesystem and gathering
+/// timeline info.
+struct RecurseState {
+    device: u64,
+    ids: Vec<ChunkDigest>,
+    encoder: GzChunkedEncoder,
+}
+
+/// Retrieves device ID from metadata.
+#[allow(unused_variables)]
+fn dev_from_metadata(metadata: &Metadata) -> u64 {
+    cfg_if! {
+        if #[cfg(target_family = "unix")] {
+            use std::os::unix::fs::MetadataExt;
+            metadata.dev()
+        } else if #[cfg(target_family = "windows")] {
+            0
+        } else {
+            compile_error!("unsupported OS family");
+        }
+    }
+}
+
+/// Tries to convert OS-dependent string to raw bytes.
+fn bytes_from_os_str(s: &OsStr) -> std::io::Result<Vec<u8>> {
+    cfg_if! {
+        if #[cfg(target_family = "unix")] {
+            use std::os::unix::ffi::OsStrExt;
+            Ok(Vec::from(s.as_bytes()))
+        } else if #[cfg(target_family = "windows")] {
+            use std::os::windows::ffi::OsStrExt;
+            use byteorder::{LittleEndian, WriteBytesExt};
+            s.encode_wide().try_fold(Vec::new(), |mut stream, ch| {
+                stream.write_u16::<LittleEndian>(ch).map(|_| stream)
+            })
+        } else {
+            compile_error!("unsupported OS family");
+        }
+    }
+}
+
+/// Converts raw bytes to OS-dependent string.
+fn os_string_from_bytes(bytes: &[u8]) -> OsString {
+    cfg_if! {
+        if #[cfg(target_family = "unix")] {
+            use std::os::unix::ffi::OsStringExt;
+            OsString::from_vec(Vec::from(bytes))
+        } else if #[cfg(target_family = "windows")] {
+            use std::os::windows::ffi::OsStringExt;
+            use byteorder::{LittleEndian, ReadBytesExt};
+            let mut wchars = Vec::new();
+            let mut bytes_slice = bytes;
+            while let Ok(wchar) = bytes_slice.read_u16::<LittleEndian>() {
+                wchars.push(wchar);
+            }
+            OsString::from_wide(wchars.as_slice())
+        } else {
+            compile_error!("unsupported OS family");
+        }
+    }
+}
+
+/// Encodes filesystem metadata into timeline entry proto.
+fn entry_from_metadata(metadata: &Metadata, path: &PathBuf) -> std::io::Result<TimelineEntry>
```

Comment:
Passing `PathBuf` by reference is an anti-pattern (same way as `&String` or `&Vec<_>`). Prefer to use `Path` instead.